dhcpcd.8
.\" $Id$
.\"
.TH DHCPCD 8 "18 Jul 2007" "dhcpcd 3.1"

.SH NAME
dhcpcd \- DHCP client daemon

.SH SYNOPSIS
.in +.5i
.ti -.5i
dhcpcd
\%[\-dknpAEGHMLNRTY]
\%[\-c\ script]
\%[\-h\ hostname]
\%[\-i\ vendorClassID]
\%[\-l\ leasetime]
\%[\-m\ metric]
\%[\-r\ ipaddr]
\%[\-s\ ipaddr[/cidr]]
\%[\-t\ timeout]
\%[\-u\ userClass]
\%[\-F\ none | ptr | both ]
\%[\-I\ clientID ]
\%[interface]
.in -.5i
.SH DESCRIPTION
.B dhcpcd
is an implementation of the DHCP client specified in
.B RFC 2131.

It gets the host information (IP address, netmask, broadcast address,
etc.) from a DHCP server and configures the network interface of the
machine on which it is running. It also tries to renew the lease time
according to
.B RFC 2131.

If
.B dhcpcd
fails to get a lease then we attempt Dynamic Configuration of IPv4
Link-Local Addresses unless the
.B \-L
option is given.

.SH OPTIONS
.TP
.BI interface
Specifies the network interface name (eth0, eth1, etc.).
.TP
.BI \-c \ script
.B dhcpcd
will try to execute
.I script
instead of the default script
.I /etc/dhcpcd.sh
every time it configures or brings down the interface. See the
description of
.I dhcpcd.sh
script in
.B FILES
section below.
.TP
.BI \-d
Echos debugging and information messages to the console.
Subsequent debug options stop \fBdhcpcd\fR from daemonising.
.TP
.BI \-h \ hostname
specifies a string used for the hostname option field when
.B dhcpcd
sends DHCP messages. Some DHCP servers, notably those used by
@Home Networks, require the hostname option
field containing a specific string in the DHCP messages from clients.
When combined with the -F switch, specifies the string used for the
FQDN option field instead of the hostname option field.
We send the current hostname by default. To send no hostname, use -h ''.
.TP
.BI \-i \ vendorClassID
Specifies the vendor class identifier string. The default is dhcpcd-<version>.
.TP
.BI \-k
Sends
.B SIGHUP
signal to the
.B dhcpcd
process associated with the specified interface if one is currently running. If
.B dhcpcd
receives
.B SIGHUP
it will send
.B DHCP_RELEASE
message to the server and destroy dhcpcd cache. In a case
.B dhcpcd
receives
.B SIGTERM
which is normally used by
.B shutdown(8)
when rebooting the system
.B dhcpcd
will not send
.B DHCP_RELEASE
and will not destroy cache. When system boots
.B dhcpcd
will use cache to request the same IP address
from DHCP server which was assigned before the
system went down. (see also
.B -p
)
.TP
.BI \-l \ leasetime
Specifies (in seconds) the recommended lease time to the server. (Note
that the server can override this value if it sees fit). This value is
used in the
.B DHCP_DISCOVER
message. Use -1 for an infinite lease time. We don't request a specific
lease time by default. If we do not receive a lease time in the
.B DHCP_OFFER
message then we default to 1 hour.
.TP
.BI \-m \ metric
Routes will be added with the given metric. The default is 0.
On some systems such as FreeBSD the interface is given the metric.
.TP
.BI \-n
Sends
.B SIGALRM
signal to the
.B dhcpcd
process that is currently running which
forces
.B dhcpcd
to try to renew the lease. If dhcpcd is not running, the flag
is ignored and
.B dhcpcd
follows the normal startup procedure.
.TP
.BI \-p
Stops
.B dhcpcd
from removing the interface configuration when it is terminated with the
.B SIGTERM
signal. This is useful when a host is running with an NFS-mounted root
filesystem over an interface controlled by DHCP. It should not be used
except in those circumstances, since if 
.B dhcp
is stopped it can no longer down an interface at the end of its
lease period when the lease is not renewed.
.TP
.BI \-r \ ipaddr
Sends
.B DHCP_REQUEST
message requesting to lease IP address ipaddr.
The ipaddr parameter must be in the form xxx.xxx.xxx.xxx.
If we are not given an ipaddr then we try to use the first one assigned
to the interface if present, otherwise the last address we leased for the
interface, otherwise we return an error.
This effectively doubles the timeout period, as if we fail to get
this IP address then we enter the INIT state and try to get any
IP address.
.TP
.BI \-s \ ipaddr[/cidr]
Exactly like \fB-r\fR, but sends DHCP_INFORM instead. This requires the
interface to be configured with the ipaddr/cidr first.
When we DHCP_INFORM, we don't request or respect any lease times.
However, we do re-inform by the lease time specified by
.BI \-l
if given.
.TP
.BI \-t \ timeout
Specifies (in seconds ) for how long
.B dhcpcd
will try to get an IP address. The default is 20 seconds.
.B dhcpcd
will not fork into background until it gets a valid IP address
in which case dhcpcd will return 0 to the parent process.
In a case
.B dhcpcd
times out before receiving a valid IP address from DHCP server
.B dhcpcd
will return exit code 1 to the parent process. Setting the timeout to
zero disables it: dhcp will keep trying forever to get a lease, and if
the lease is lost, it will try forever to get another.
.TP
.BI \-u \ userClass
Tags the  DHCP message with the specified user class. DHCP servers can use
these fields to send back different information instead of grouping by
fixed hardware addresses. You can specify more than one user class, but the
total length must be less than 255 characters, -1 character for each user
class.
.TP
.BI \-x
Sends
.B SIGTERM
signal to the
.B dhcpcd
process associated with the specified interface if one is currently running. If
.B dhcpcd
receives
.B SIGTERM
it will stop running, de-configure the interface and exit.
(see also
.B -k
and
.B -p
)
.TP
.BI \-A
Don't do an
.B ARP
check on the IP address. 
.TP
.BI \-E
Will read 
.I /var/lib/dhcpcd/dhcpcd-<interface>.info
file and use last known good lease if
.B dhcpcd
is unable to reach the DHCP server and the lease has not expired.
This puts \fBdhcpcd\fR into the BOUND or REBINDING state depending on
how long the lease has until expiry.
.TP
.BI \-F \ none | ptr | both
Forces
.B dhcpcd
to request the DHCP server update the DNS using the FQDN option
instead of the Hostname option. The name used by this option
is specified with the \fB-h\fP switch, which must be present. If
the \fB-h\fP switch is not present, the FQDN option is ignored.
The name should be fully qualified, although servers usually
accept a simple name.
.I both
requests that the DHCP server update both the A and PTR
records in the DNS.
.I ptr
requests that the DHCP server updates only the PTR record in
the DNS.
.I none
requests that the DHCP server perform no updates.
.B dhcpcd
does not perform any DNS update, even when the server is
requested to perform no updates.  This can be easily
implemented outside the client; all the necessary
information is recorded in the
.I /var/lib/dhcpcd/dhcpcd-<interface>.info
file.
.TP
.BI \-G
Prevents
.B dhcpcd
from installing default routes provided by DHCP server.
.TP
.BI \-H
Forces
.B dhcpcd
to set hostname of the host to the hostname option supplied by DHCP server.
By default
.B dhcpcd
will NOT set hostname of the host to the hostname option
received from DHCP server unless the current hostname is blank, (none) or
localhost. If no hostname is returned by the DHCP server then we attempt
to lookup the hostname via DNS.
More -H's control what we do with the FQDN returned by DNS.
.IP
.BI \-H
set hostname to the full FQDN
.br
.BI \-HH
strip the domain if it matches a given domain in our DHCP message
.br
.BI \-HHH
strip the domain regardless
.br
.BI \-HHHH
force hostname lookup even if given a hostname in our DHCP message
.br
.BI \-HHHHH
same as above, but strip the domain if it matches
.br
.BI \-HHHHHH
same as above, but strip the domain regardless
.TP
.BI \-I \ clientID
Specifies the client identifier string. If not specified then
.B dhcpcd
will attempt to create a client identifier according to \fBRFC 4361\fR
and store the DUID part in
.I /var/lib/dhcpcd/dhcpcd.duid\fR, otherwise
.B dhcpcd
uses the MAC address of the network interface. If \fB-I\fR is not given
an option then we use the MAC address of the network interface.
.TP
.BI \-L
Prevents dhcpcd from probing for IPV4LL addresses. IPV4LL is otherwise
known as ZeroConf or APIPA and is \fBRFC 3927\fR.
.TP
.BI \-M
Prevents
.B dhcpcd
from setting the \fIMTU\fR provided by the DHCP server.
.TP
.BI \-N
Prevents
.B dhcpcd
from replacing
.I /etc/ntp.conf
.TP
.BI \-R
Prevents
.B dhcpcd
from replacing
.I /etc/resolv.conf
or using resolvconf.
.TP
.BI \-T
dhcpcd sends out a DHCP_DISCOVER message and then prints the values returned to
stdout. It does not configure the interface or touch the .info files.
.TP
.BI \-Y
Prevents
.B dhcpcd
from replacing
.I /etc/yp.conf

.SH FILES
.PD 0
.TP
.BI /var/lib/dhcpcd
Directory used for storing files information files created by
.B dhcpcd
that can be used by shell scripts.
.PD 1
.TP
.BI /etc/dhcpcd.sh
script file, which
.B dhcpcd
will try to execute whenever it configures or brings down the interface. The
path to this executable script can be changed with
.I \-c \ script 
option.
.B dhcpcd
passes 3 parameters to
.I dhcpcd.sh
script:
.TP
.I dhcpcd.sh infofile [up | down | new]
The first parameter infofile is the path to a file containing all DHCP
information we have. The second parameter value
.I up | down | new
mean the interface has been brought up with the same IP address as before ("up"), or
with the new IP address ("new"), or the interface has been brought down ("down").
.TP
.BI /etc/resolv.conf
file created by
.B dhcpcd
when the client receives DNS and domain name options.
If resolvconf is present on the system then we send the data to it instead
of overwriting resolv.conf
.TP
.BI /etc/yp.conf
file created by
.B dhcpcd
when the client receives NIS options.
.TP
.BI /etc/ntp.conf
file created by
.B dhcpcd
when the client receives NTP options.
.TP
.BI /var/run/dhcpcd-<interface>.pid
file containing the process id of
.B dhcpcd.
The word
.I <interface>
is actually replaced with the network interface name like
.I eth0
to which
.B dhcpcd
is attached.

.SH SEE ALSO
.LP
.I Dynamic Host Configuration Protocol,
RFC 2131
.LP
.I DHCP Options and BOOTP Vendor Extensions,
RFC 2132
.LP
.I Dynamic Configuration of IPv4 Link-Local Addresses, 
RFC 3927
.LP
.I DHCP FQDN Option specification,
RFC 4702

.SH BUGS
Please report them to http://dhcpcd.berlios.de or http://bugs.gentoo.org.
.PD 0

.SH AUTHORS
Roy Marples <roy@marples.name>


ChangeLog
Work with the BSD RC system too.
Find resolvconf in path instead of /sbin.
make HAVE_FORK=yes|no should be used if cross compiling.
Implement an exponential NAK backoff.
Clear the signal_pipe fd from the fdset when all signals have
been read.
Don't release a link local address.
Use sysconf to detect if we have a working monotonic clock.
Only request NTP, NIS, etc if we have compiled that feature in.
dhcpcd --version now shows what compile time options were used.

dhcpcd-3.1.7
Add warnings about requesting a link local address.
Always add a link local route unless we're given a non-private
address by the DHCP server.
Reduce stack usage by using malloc more.
If we're on a different subnet from the one we get DHCP for, don't
use our current address in messages. Thanks to Wilson Callan.
Skip over bogus EINTR error on select when arp checking for a
different address from what we already have.

dhcpcd-3.1.6
--skiproutes has been added for no fork users so that we know if we
can delete the route at a later date.
Many thanks to Michael Durrant for testing the below changes.
Compile and work on OSX/Darwin.
If we have no fork then we re-exec ourselves with --daemonised.
Improved the build system so we automatically detect if we need
librt or libresolv and if fork() works or not.
Move back to using librt for clock_gettime on Linux systems.

dhcpcd-3.1.5
Fix the flushing of addresses on BSD systems.
Rework get_time so that we don't link to librt on Linux.
Thanks to regenrecht for reporting the below three overflows:-
 Fix a potential buffer overflow in hwaddr_ntoa if length > 42.
 Fix a potential heap overflow in decode_CSR when CIDR > 32.
 Fix a potential heap overflow in decode_sipservers when length is 1.
-L actually doesn't do IPv4LL as advertised.

dhcpcd-3.1.4
Ensure ARP checking times out when there is a flood.
Add -x option to quit without releasing the lease.

dhcpcd-3.1.3
Fix link routes which broke in 3.1.2.

dhcpcd-3.1.2
Use a generic route logger.
Fix static route netmask calculation and applicaton.
Use --nodeps when restarting services.
Simply CIDR calculation, thanks to Francois-Xavier Le Bail.
Don't free the dhcp object on RENEW (fixes a segfault).

dhcpcd-3.1.1
Fix segfault in arp code that some users had.
Remove getifaddrs code and instead just use ioctls so we don't break
platforms without getifaddrs as often.

dhcpcd-3.1.0
Add new socket code to try and support InfiniBand better.
Support RFC 3361 SIP Servers.
We now do ARP checking by default as recommended by RFC 2131.
Add RFC 3927 (aka IPV4LL aka APIPA) support by default.
Suport DHCP option (52) overload.
Added -T test option. This just sends a DHCP_DISCOVER message and then
prints the configuration data to stdout - we don't request a lease,
configure the interface or write the info file.
Fix adding of routes on the BSDs.
ROOTPATH should work correctly again, thanks to danhien.
We now have 6 hostname options, which should satisfy everyone.
Fix arp compilation warning on GCC-4.2.0.
Enable DHCP_INFORM support via the -s option again, moving DHCP_REQUEST
to the -r option. This makes us more compatible with dhcpcd-1.x and 2.x.
Ensure that dhcpcd.sh is always called.
flock the pidfile and stop writing the parent pid to it so we're more robst.
Use the DUID stored to create an RFC 4361 conformant client identifier
if none is specified instead of just using the MAC address.
Create a DUID-LLT according to RFC 3315 and store it in peristent file.
config.h now has the compile time options, instead of being dotted around.
Added -E option, which reads the last info file and uses the information there
if we timeout and the lease is still valid, thanks to Roberto Angelino.
Compiles ok on NetBSD thanks to Gabor Z. Papp.
If the current hostname contains dots then send it as a FQDN.
Ensure that static routes are always added before routers.
Use getnameinfo instead of gethostbyaddr.
Remove gateways from the ROUTES and add to GATEWAYS
define #INFO_COMPAT if you wish your .info file to add support for some
old 1.x and 2.x structure.

dhcpcd-3.0.17
Minimum message size is now 300 bytes for crappy DHCP servers who think
they have to obey the BOOTP minimum message size :/
Compile ok on OpenBSD (thanks to icezimm and reb in #gentoo-bsd for testing)
We check gcc supports -Wextra and friends before using them.
We now restore the starting MTU value when we exit OR we don't receive a
a valid MTU value from the dhcp server.
--nomtu / -M now stops the MTU from being used.
You can now build dhcpcd without support for writing ntp, nis and info files
which makes the binary a few k smaller.
Support OpenNTP as well as NTP.
Re-work the state engine a little, so we renew and rebind correctly.
Subsequent debug options stop dhcpcd from daemonising.
When we get an invalid length for a DHCP option, try and continue anyway.
When MTU is less than 576 we now ignore it instead of setting the MTU to 576.
Build correctly on dietlibc, thanks to d00mer.
Check that if_nametoindex returns a valid index, thanks to d00mer.

dhcpcd-3.0.16
RFC 2131 is full of confusion regarding MTU it seems as the effective minimum
MTU is really 576 or DHCP requests stop working.

dhcpcd-3.0.15
Fix MTU length checking.
We now request MTU from the DHCP server, and if given we set it.

dhcpcd-3.0.14
Don't crash with 0 or invalid length DHCP options, reported by
Stefan de Konink.

dhcpcd-3.0.13
Fix regression on Linux for sending packets over non Ethernet devices.
define ARPHRD_IEEE1394 if it doesn not exist, like for Linux-2.4 kernels.

dhcpcd-3.0.12
Test to see if the pid in the pidfile is actually running before claiming
that we are.
Fix the `netlink message truncated` error since 3.0.6.
We now ensure that our renewal and rebind times are sane.
Fix flushing interfaces for systems without getifaddrs and >1 interface.
All netlink error messages are now prefixed with netlink.
Add DHCP over FireWire 1394 (RFC 2855) and InfiniBand (RFC 4390) support.
We should add support for DUIDs (RFC 4361) as well.

dhcpcd.c
/*
 * dhcpcd - DHCP client daemon
 * Copyright 2006-2007 Roy Marples <roy@marples.name>
 * 
 * Distributed under the terms of the GNU General Public License v2
 */

const char copyright[] = "Copyright (c) 2006-2007 Roy Marples";

/* We need to define this to get kill on GNU systems */
#ifdef __linux__
#define _BSD_SOURCE
#define _POSIX_SOURCE
#endif

#include <sys/file.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <arpa/inet.h>
#include <errno.h>
#include <fcntl.h>
#include <getopt.h>
#include <paths.h>
#include <signal.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include "config.h"
#include "client.h"
#include "dhcpcd.h"
#include "dhcp.h"
#include "interface.h"
#include "logger.h"
#include "version.h"

static int doversion = 0;
static int dohelp = 0;
#define EXTRA_OPTS
static const struct option longopts[] = {
	{"arp",         no_argument,        NULL, 'a'},
	{"script",      required_argument,  NULL, 'c'},
	{"debug",       no_argument,        NULL, 'd'},
	{"hostname",    optional_argument,  NULL, 'h'},
	{"classid",     optional_argument,  NULL, 'i'},
	{"release",     no_argument,        NULL, 'k'},
	{"leasetime",   required_argument,  NULL, 'l'},
	{"metric",      required_argument,  NULL, 'm'},
	{"renew",       no_argument,        NULL, 'n'},
	{"persistent",  no_argument,        NULL, 'p'},
	{"inform",      optional_argument,  NULL, 's'},
	{"request",     optional_argument,  NULL, 'r'},
	{"timeout",     required_argument,  NULL, 't'},
	{"userclass",   required_argument,  NULL, 'u'},
	{"exit",        no_argument,        NULL, 'x'},
	{"lastlease",   no_argument,        NULL, 'E'},
	{"fqdn",        required_argument,  NULL, 'F'},
	{"nogateway",   no_argument,        NULL, 'G'},
	{"sethostname", no_argument,        NULL, 'H'},
	{"clientid",    optional_argument,  NULL, 'I'},
	{"noipv4ll",	no_argument,		NULL, 'L'},
	{"nomtu",       no_argument,        NULL, 'M'},
	{"nontp",       no_argument,        NULL, 'N'},
	{"nodns",       no_argument,        NULL, 'R'},
	{"test",        no_argument,        NULL, 'T'},
	{"nonis",       no_argument,        NULL, 'Y'},
	{"help",        no_argument,        &dohelp, 1},
	{"version",     no_argument,        &doversion, 1},
#ifdef THERE_IS_NO_FORK
	{"daemonised",	no_argument,		NULL, 'f'},
	{"skiproutes",  required_argument,	NULL, 'g'},
#endif
	{NULL,          0,                  NULL, 0}
};


#define STRINGINT(_string, _int) { \
	char *_tmp; \
	long _number = strtol (_string, &_tmp, 0); \
	errno = 0; \
	if ((errno != 0 && _number == 0) || _string == _tmp || \
		(errno == ERANGE && (_number == LONG_MAX || _number == LONG_MIN))) \
	{ \
		logger (LOG_ERR, "`%s' out of range", _string);; \
		exit (EXIT_FAILURE); \
	} \
	else \
	_int = (int) _number; \
}

#ifdef THERE_IS_NO_FORK
char dhcpcd[PATH_MAX];
char **dhcpcd_argv = NULL;
int dhcpcd_argc = 0;
char *dhcpcd_skiproutes = NULL;
#undef EXTRA_OPTS
#define EXTRA_OPTS "fg:"
#endif

static pid_t read_pid (const char *pidfile)
{
	FILE *fp;
	pid_t pid = 0;

	if ((fp = fopen (pidfile, "r")) == NULL) {
		errno = ENOENT;
		return 0;
	}

	fscanf (fp, "%d", &pid);
	fclose (fp);

	return pid;
}

static void usage ()
{
	printf ("usage: "PACKAGE" [-adknpEGHMNRTY] [-c script] [-h hostame] [-i classID]\n"
	        "              [-l leasetime] [-m metric] [-r ipaddress] [-s ipaddress]\n"
			"              [-t timeout] [-u userclass] [-F none | ptr | both]\n"
			"              [-I clientID] <interface>\n");
}

int main(int argc, char **argv)
{
	options_t *options;
	int userclasses = 0;
	int opt;
	int option_index = 0;
	char *prefix;
	pid_t pid;
	int debug = 0;
	int i;
	int pidfd = -1;
	int sig = 0;

	/* Close any un-needed fd's */
	for (i = getdtablesize() - 1; i >= 3; --i)
		close (i);

	openlog (PACKAGE, LOG_PID, LOG_LOCAL0);

	options = xmalloc (sizeof (options_t));
	memset (options, 0, sizeof (options_t));
	options->script = (char *) DEFAULT_SCRIPT;
	snprintf (options->classid, CLASS_ID_MAX_LEN, "%s %s", PACKAGE, VERSION);
	options->classid_len = strlen (options->classid);

	options->doarp = true;
	options->dodns = true;
	options->domtu = true;
	options->donis = true;
	options->dontp = true;
	options->dogateway = true;
	options->daemonise = true;
	options->doinform = false;
	options->doipv4ll = true;
	options->timeout = DEFAULT_TIMEOUT;

	gethostname (options->hostname, sizeof (options->hostname));
	if (strcmp (options->hostname, "(none)") == 0 ||
		strcmp (options->hostname, "localhost") == 0)
		memset (options->hostname, 0, sizeof (options->hostname));

	/* Don't set any optional arguments here so we retain POSIX
	 * compatibility with getopt */
	while ((opt = getopt_long(argc, argv, EXTRA_OPTS
							  "c:dh:i:kl:m:npr:s:t:u:xAEF:GHI:LMNRTY",
							  longopts, &option_index)) != -1)
	{
		switch (opt) {
			case 0:
				if (longopts[option_index].flag)
					break;
				logger (LOG_ERR, "option `%s' should set a flag",
						longopts[option_index].name);
				exit (EXIT_FAILURE);
				break;
			case 'c':
				options->script = optarg;
				break;
			case 'd':
				debug++;
				switch (debug) {
					case 1:
						setloglevel (LOG_DEBUG);
						break;
					case 2:
						options->daemonise = false;
						break;
				}
				break;
#ifdef THERE_IS_NO_FORK
			case 'f':
				options->daemonised = true;
				close_fds ();
				break;
			case 'g':
				dhcpcd_skiproutes = xstrdup (optarg);
				break;
#endif
			case 'h':
				if (! optarg)
					memset (options->hostname, 0, sizeof (options->hostname));
				else if (strlen (optarg) > MAXHOSTNAMELEN) {
					logger (LOG_ERR, "`%s' too long for HostName string, max is %d",
							optarg, MAXHOSTNAMELEN);
					exit (EXIT_FAILURE);
				} else
					strlcpy (options->hostname, optarg, sizeof (options->hostname));
				break;
			case 'i':
				if (! optarg) {
					memset (options->classid, 0, sizeof (options->classid));
					options->classid_len = 0;
				} else if (strlen (optarg) > CLASS_ID_MAX_LEN) {
					logger (LOG_ERR, "`%s' too long for ClassID string, max is %d",
							optarg, CLASS_ID_MAX_LEN);
					exit (EXIT_FAILURE);
				} else
					options->classid_len = strlcpy (options->classid, optarg,
												   sizeof (options->classid));
				break;
			case 'k':
				sig = SIGHUP;
				break;
			case 'l':
				STRINGINT (optarg, options->leasetime);
				if (options->leasetime <= 0) {
					logger (LOG_ERR, "leasetime must be a positive value");
					exit (EXIT_FAILURE);
				}
				break;
			case 'm':
				STRINGINT (optarg, options->metric);
				break;
			case 'n':
				sig = SIGALRM;
				break;
			case 'p':
				options->persistent = true;
				break;
			case 's':
				options->doinform = true;
				options->doarp = false;
				if (! optarg || strlen (optarg) == 0) {
					options->request_address.s_addr = 0;
					break;
				} else {
					char *slash = strchr (optarg, '/');
					if (slash) {
						int cidr;
						/* nullify the slash, so the -r option can read the
						 * address */
						*slash++ = '\0';
						if (sscanf (slash, "%d", &cidr) != 1 ||
							inet_cidrtoaddr (cidr, &options->request_netmask) != 0) {
							logger (LOG_ERR, "`%s' is not a valid CIDR", slash);
							exit (EXIT_FAILURE);
						}
					}
					/* fall through */
				}
			case 'r':
				if (! options->doinform)
					options->dorequest = true;
				if (strlen (optarg) > 0 &&
					! inet_aton (optarg, &options->request_address))
				{ 
					logger (LOG_ERR, "`%s' is not a valid IP address", optarg);
					exit (EXIT_FAILURE);
				}
				break;
			case 't':
				STRINGINT (optarg, options->timeout);
				if (options->timeout < 0) {
					logger (LOG_ERR, "timeout must be a positive value");
					exit (EXIT_FAILURE);
				}
				break;
			case 'u':
				{
					int offset = 0;
					for (i = 0; i < userclasses; i++)
						offset += (int) options->userclass[offset] + 1;
					if (offset + 1 + strlen (optarg) > USERCLASS_MAX_LEN) {
						logger (LOG_ERR, "userclass overrun, max is %d",
								USERCLASS_MAX_LEN);
						exit (EXIT_FAILURE);
					}
					userclasses++;
					memcpy (options->userclass + offset + 1 , optarg, strlen (optarg));
					options->userclass[offset] = strlen (optarg);
					options->userclass_len += (strlen (optarg)) + 1;
				}
				break;
			case 'x':
				sig = SIGTERM;
				break;
			case 'A':
#ifndef ENABLE_ARP
				logger (LOG_ERR, "arp support not compiled into dhcpcd");
				exit (EXIT_FAILURE);
#endif
				options->doarp = false;
				break;
			case 'E':
#ifndef ENABLE_INFO
				logger (LOG_ERR, "info support not compiled into dhcpcd");
				exit (EXIT_FAILURE);
#endif
				options->dolastlease = true;
				break;
			case 'F':
				if (strncmp (optarg, "none", strlen (optarg)) == 0)
					options->fqdn = FQDN_NONE;
				else if (strncmp (optarg, "ptr", strlen (optarg)) == 0)
					options->fqdn = FQDN_PTR;
				else if (strncmp (optarg, "both", strlen (optarg)) == 0)
					options->fqdn = FQDN_BOTH;
				else {
					logger (LOG_ERR, "invalid value `%s' for FQDN", optarg);
					exit (EXIT_FAILURE);
				}
				break;
			case 'G':
				options->dogateway = false;
				break;
			case 'H':
				options->dohostname++;
				break;
			case 'I':
				if (optarg) {
					if (strlen (optarg) > CLIENT_ID_MAX_LEN) {
						logger (LOG_ERR, "`%s' is too long for ClientID, max is %d",
								optarg, CLIENT_ID_MAX_LEN);
						exit (EXIT_FAILURE);
					}
					options->clientid_len = strlcpy (options->clientid, optarg,
													sizeof (options->clientid));
					/* empty string disabled duid */
					if (options->clientid_len == 0)
						options->clientid_len = -1;
				} else {
					memset (options->clientid, 0, sizeof (options->clientid));
					options->clientid_len = -1;
				}
				break;
			case 'L':
				options->doipv4ll = false;
				break;
			case 'M':
				options->domtu = false;
				break;
			case 'N':
				options->dontp = false;
				break;
			case 'R':
				options->dodns = false;
				break;
			case 'T':
#ifndef ENABLE_INFO
				logger (LOG_ERR, "info support not compiled into dhcpcd");
				exit (EXIT_FAILURE);
#endif
				options->test = true;
				options->persistent = true;
				break;
			case 'Y':
				options->donis = false;
				break;
			case '?':
				usage ();
				exit (EXIT_FAILURE);
			default:
				usage ();
				exit (EXIT_FAILURE);
		}
	}
	if (doversion) {
		printf (""PACKAGE" "VERSION"\n");
		printf ("Compile time options:"
#ifdef ENABLE_ARP
				" ARP"
#endif
#ifdef ENABLE_DUID
				" DUID"
#endif
#ifdef ENABLE_INFO
				" INFO"
#endif
#ifdef ENABLE_INFO_COMPAT
				" INFO_COMPAT"
#endif
#ifdef ENABLE_IPV4LL
				" IPV4LL"
#endif
#ifdef ENABLE_NIS
				" NIS"
#endif
#ifdef ENABLE_NTP
				" NTP"
#endif
#ifdef SERVICE
				" " SERVICE
#endif
#ifdef ENABLE_RESOLVCONF
				" RESOLVCONF"
#endif
#ifdef THERE_IS_NO_FORK
				" THERE_IS_NO_FORK"
#endif
				"\n");
	}

	if (dohelp)
		usage ();

#ifdef THERE_IS_NO_FORK
	dhcpcd_argv = argv;
	dhcpcd_argc = argc;
	if (! realpath (argv[0], dhcpcd)) {
		logger (LOG_ERR, "unable to resolve the path `%s': %s",
				argv[0], strerror (errno));
		exit (EXIT_FAILURE);
	}
#endif

	if (optind < argc) {
		if (strlen (argv[optind]) > IF_NAMESIZE) {
			logger (LOG_ERR, "`%s' is too long for an interface name (max=%d)",
					argv[optind], IF_NAMESIZE);
			exit (EXIT_FAILURE);
		}
		strlcpy (options->interface, argv[optind],
				 sizeof (options->interface));
	} else {
		/* If only version was requested then exit now */
		if (doversion || dohelp)
			exit (EXIT_SUCCESS);

		logger (LOG_ERR, "no interface specified");
		exit (EXIT_FAILURE);
	}

	if (strchr (options->hostname, '.')) {
		if (options->fqdn == FQDN_DISABLE)
			options->fqdn = FQDN_BOTH;
	} else
		options->fqdn = FQDN_DISABLE;

	if (options->request_address.s_addr == 0 && options->doinform) {
		if ((options->request_address.s_addr = get_address (options->interface)) != 0)
			options->keep_address = true;
	}

	if (IN_LINKLOCAL (ntohl (options->request_address.s_addr))) {
		logger (LOG_ERR, "you are not allowed to request a link local address");
		exit (EXIT_FAILURE);
	}

	if (geteuid ()) {
		logger (LOG_ERR, "you need to be root to run "PACKAGE);
		exit (EXIT_FAILURE);
	}

	prefix = xmalloc (sizeof (char) * (IF_NAMESIZE + 3));
	snprintf (prefix, IF_NAMESIZE, "%s: ", options->interface);
	setlogprefix (prefix);
	snprintf (options->pidfile, sizeof (options->pidfile), PIDFILE,
			  options->interface);
	free (prefix);

	chdir ("/");
	umask (022);
	
	if (mkdir (CONFIGDIR, S_IRUSR |S_IWUSR |S_IXUSR | S_IRGRP | S_IXGRP
			   | S_IROTH | S_IXOTH) && errno != EEXIST )
	{
		logger (LOG_ERR, "mkdir(\"%s\",0): %s\n", CONFIGDIR, strerror (errno));
		exit (EXIT_FAILURE);
	}

	if (mkdir (ETCDIR, S_IRUSR | S_IWUSR | S_IXUSR | S_IRGRP | S_IXGRP
			   | S_IROTH | S_IXOTH) && errno != EEXIST )
	{
		logger (LOG_ERR, "mkdir(\"%s\",0): %s\n", ETCDIR, strerror (errno));
		exit (EXIT_FAILURE);
	}

	if (options->test) {
		if (options->dorequest || options->doinform) {
			logger (LOG_ERR, "cannot test with --inform or --request");
			exit (EXIT_FAILURE);
		}

		if (options->dolastlease) {
			logger (LOG_ERR, "cannot test with --lastlease");
			exit (EXIT_FAILURE);
		}

		if (sig != 0) {
			logger (LOG_ERR, "cannot test with --release or --renew");
			exit (EXIT_FAILURE);
		}
	}

	if (sig != 0) {
		int killed = -1;
		pid = read_pid (options->pidfile);
		if (pid != 0)
			logger (LOG_INFO, "sending signal %d to pid %d", sig, pid);

		if (! pid || (killed = kill (pid, sig)))
			logger (sig == SIGALRM ? LOG_INFO : LOG_ERR, ""PACKAGE" not running");

		if (pid != 0 && (sig != SIGALRM || killed != 0))
			unlink (options->pidfile);

		if (killed == 0)
			exit (EXIT_SUCCESS);

		if (sig != SIGALRM)
			exit (EXIT_FAILURE);
	}

	if (! options->test && ! options->daemonised) {
		if ((pid = read_pid (options->pidfile)) > 0 && kill (pid, 0) == 0) {
			logger (LOG_ERR, ""PACKAGE" already running on pid %d (%s)",
					pid, options->pidfile);
			exit (EXIT_FAILURE);
		}

		pidfd = open (options->pidfile, O_WRONLY | O_CREAT | O_NONBLOCK, 0660);
		if (pidfd == -1) {
			logger (LOG_ERR, "open `%s': %s", options->pidfile, strerror (errno));
			exit (EXIT_FAILURE);
		}

		/* Lock the file so that only one instance of dhcpcd runs on an interface */
		if (flock (pidfd, LOCK_EX | LOCK_NB) == -1) {
			logger (LOG_ERR, "flock `%s': %s", options->pidfile, strerror (errno));
			exit (EXIT_FAILURE);
		}

		/* dhcpcd.sh should not interhit this fd */
		if ((i = fcntl (pidfd, F_GETFD, 0)) == -1 ||
			fcntl (pidfd, F_SETFD, i | FD_CLOEXEC) == -1)
			logger (LOG_ERR, "fcntl: %s", strerror (errno));

		writepid (pidfd, getpid ());
		logger (LOG_INFO, PACKAGE " " VERSION " starting");
	}

	/* Seed random */
	srandomdev ();

	i = EXIT_FAILURE;
	if (dhcp_run (options, &pidfd) == 0)
		i = EXIT_SUCCESS;

	/* If we didn't daemonise then we need to punt the pidfile now */
	if (pidfd > -1) {
		close (pidfd);
		unlink (options->pidfile);
	}

	free (options);

#ifdef THERE_IS_NO_FORK
	/* There may have been an error before the dhcp_run function
	 * clears this, so just do it here to be safe */
	free (dhcpcd_skiproutes);
#endif

	logger (LOG_INFO, "exiting");
	
	exit (i);
}
